### PR TITLE
Optimize: add dedicated BeaconSate cache container, add active validators cache

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description
* Add `TransitionCaches` instance to the `BeaconStateWithCaches`. It would contain all the caches persistent across transitions. The `TransitionCaches` should be 
  * thread-safe: it should safe to parallelize any spec helper function with the same `TransitionCaches` instance
  * fork-safe: any cache modification made in one fork shouldn't be visible in another fork
* Add active validators cache. `get_active_validator_indices` is frequently invoked while its return value is persistent within an epoch.

## Results

**Note**: below results are not absolute (may differ from results measured on this branch) and should be treated relatively.

#### Before
```
Benchmark                              (validatorsCount)  Mode  Cnt   Score   Error  Units
TransitionBenchmark.Block.importBlock               1024    ss   50   0,352 ± 0,004   s/op
TransitionBenchmark.Block.importBlock               3072    ss   50   0,527 ± 0,003   s/op
TransitionBenchmark.Epoch.importBlock               1024    ss   10   2,053 ± 0,035   s/op
TransitionBenchmark.Epoch.importBlock               3072    ss   10  16,164 ± 0,146   s/op
```
#### After
```
Benchmark                              (validatorsCount)  Mode  Cnt  Score   Error  Units
TransitionBenchmark.Block.importBlock               1024    ss   50  0,351 ± 0,009   s/op
TransitionBenchmark.Block.importBlock               3072    ss   50  0,529 ± 0,005   s/op
TransitionBenchmark.Block.importBlock              10240    ss   50  1,135 ± 0,020   s/op
TransitionBenchmark.Epoch.importBlock               1024    ss   10  0,613 ± 0,011   s/op
TransitionBenchmark.Epoch.importBlock               3072    ss   10  2,704 ± 0,090   s/op
TransitionBenchmark.Epoch.importBlock              10240    ss   10 28,404 ± 3,693   s/op
```

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  public LRUCache(int capacity) {
+    this(capacity, Collections.emptyMap());
+  }
+
+  private LRUCache(int capacity, Map<K, V> initialCachedContent) {
+    this.maxCapacity = capacity;
+    this.cacheData =
+        Collections.synchronizedMap(
+            new LinkedHashMap<>(maxCapacity + 1) {
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java]: Could this just use `LimitedHashMap`?  The static block to add initialCachedContent would just have to be done after construction but that should be safe since we're still in the constructor so no other operations can be happening.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.function.Function;
+
+/**
+ * Cache made around LRU-map with fixed size, removing eldest entries (by added) when the space is
+ * over
+ *
+ * @param <K> Keys type
+ * @param <V> Values type
+ */
+public class LRUCache<K, V> implements Cache<K, V> {
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:29]: I was going to suggest we consider using the Guava cache libraries rather than writing our own, but their version of `Cache.get` throws a checked exception to allow the fallback function to throw checked exceptions which makes using the cache kind of awful.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java [hunk 3]
```diff
@@ -100,11 +101,17 @@ public static boolean is_eligible_for_activation(BeaconState state, Validator va
...
+        .getActiveValidators()
+        .get(
+            epoch,
+            e -> {
+              List<Validator> validators = state.getValidators();
+              return IntStream.range(0, validators.size())
+                  .filter(index -> is_active_validator(validators.get(index), epoch))
+                  .boxed()
+                  .collect(Collectors.toList());
+            });
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java:114]: This works and I like how it winds up with very minimal changes, but I wonder if as we introduce more caches we'll wind up with a lot of boilerplate in getting the caches etc obscuring the actual business logic.

I think we should land this as-is, but I am wondering if we could be a bit radical here and take advantage of polymorphism.  https://github.com/PegaSysEng/artemis/commit/1144a564f684cd8d47c5d09a1323a83ab1a700e3 is what I have in mind.  If `BeaconState` provided a `getActiveValidatorIndices` method with the actual spec logic, then `BeaconStateWithCache` could just override that method to check the cache first.  Feels really nice and clean in many ways.

But there's a couple of downsides:
1. Not sure this scales well - we could wind up with half the transition code in `BeaconState`.
2. I don't see this playing well with hard forks after launch (neither does our current approach)

Which I think is enough to talk me out of it, and that it's better for classes like `ValidatorUtils` to use actual instances rather than all static methods - then this polymorphism approach could be used there, and different implementations could be provided for different hard forks, but that's a bigger scope of work.

You can safely ignore this for now if you like, I just keep pondering how we need to evolve the design to be more flexible in the future...  Happy to hear any thoughts anyone has though.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java [hunk 4]
```diff
@@ -172,6 +225,16 @@ public static BeaconStateWithCache fromBeaconState(BeaconState state) {
...
 
+  public static TransitionCaches getTransitionCaches(BeaconState state) {
+    return state instanceof BeaconStateWithCache
+        ? ((BeaconStateWithCache) state).getTransitionCaches()
+        : TransitionCaches.getNoOp();
+  }
+
+  public TransitionCaches getTransitionCaches() {
+    return transitionCaches;
+  }
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java:236]: Would it be better to add `getTransitionCaches` to `BeaconState` and then use polymorphism so `BeaconState` returns `TransitionCaches.getNoOp()` and `BeaconStateWithCache` overrides it to return a real cache?

That would avoid the `instanceof` checks here and in deepCopy.

**Review by ajsutton — APPROVED** [review]
LGTM. Couple of minor suggestions for you to consider and a bit of a ramble which didn't go anywhere. :)

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java (hunk 2 above)
Comment (Nashatyrev) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:29]: Heh, looks like we are on the same camp regarding checked exceptions :)
Never used Guava caches, but we may consider them for more advanced use cases. This one is pretty simple and straightforward.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java (hunk 1 above)
Comment (Nashatyrev) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java]: Yep 👍 
Done: https://github.com/PegaSysEng/artemis/pull/1111/commits/19d72bc794a3046bed74b5f9380a3bff80835604

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java (hunk 3 above)
Comment (Nashatyrev) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java:114]: Yep, I see what you mean and totally support this approach. The spec procedural style contradicts Java OOP. But we are still sticking to it to be closer to the original spec to easy follow spec changes. 
I think at some point this is absolutely must have, but suppose this should be a separate large refactoring and we could start it from your suggested change

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java (hunk 4 above)
Comment (Nashatyrev) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java:236]: Not sure about this. I would keep the core `BeaconState` as clear as possible of any implementation specific stuff. I believe this central class would finally become pretty complex even without those 'tool' methods. 
To reduce boilerplate in the spec functions we could do like this https://github.com/Nashatyrev/artemis/pull/6 (though after formatting I'm not sure it looks much nicer)
However I have no strong opinion whether it should done this or that way.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java (hunk 2 above)
Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:29]: Just wondering, why do you guys not like checked exceptions? To be able to handle a wider array of exceptions at a higher level?

**Review by cemozerr — APPROVED** [review]

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java (hunk 3 above)
Comment (ajsutton) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java:114]: Yeah agreed, out of scope for this PR.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java (hunk 4 above)
Comment (ajsutton) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java:236]: I'd stick with the current approach in this PR for now then.  https://github.com/Nashatyrev/artemis/pull/6 doesn't seem too worth it.

Let's see how it goes as we add more caches - we may come up with other ideas, especially if we do wind up moving to a more OO style as discussed above.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java (hunk 2 above)
Comment (ajsutton) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:29]: Checked exceptions can wind up requiring a lot of boilerplate try/catch code that makes it hard to follow control flow well.  If there's an error condition that really does need to be handled (e.g. block failing to import) then my first preference would be return a success or error result but if that's not feasible for some reason then I would use a checked exception. I wouldn't want to use an unchecked exception for something that really needs to be handled (the fact that most of the spec code throws unchecked exceptions when it fails is slightly terrifying).

Things change a bit when working with `CompletableFuture` because unchecked exceptions get picked up as failed results so well and checked exceptions are even more annoying since methods like `thenApply` take a function and don't allow throwing checked exceptions.

So it's one of those highly context dependent rules...

The most annoying thing about checked exceptions though is when a method declares a checked exception that it really shouldn't ever throw.  Which is what happens with guava's cache `get` method.  It declares a checked exception so that it can allow the fallback function to throw a checked exception, but our *actual* fallback function shouldn't ever throw any exception so we wind up being forced to handle a checked exception that shouldn't ever be thrown so there's really nothing useful we can do once we catch it. Lots of boiler plate code and no actual value.

Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:29]: That's interesting and great to know. Thank you very much for the detailed explanation @ajsutton

Comment (Nashatyrev) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:29]: @cemozerr 
From my personal experience the major part of checked exception handlings looks like 
```java
catch (SomeCheckedException e) {
  throw new SomeUncheckedException(e);
}
```
In the best case it's 
```java
catch (SomeCheckedException e) {
  throw new SomeUncheckedException("Error while doing that: " + that, e);
}
```

This is because the major part of exceptions are actually unrecoverable at a lower level from one side and from another side it doesn't make too much sense when a higher level method declares `throws` a bunch of irrelevant checked exceptions. 
Agree with @ajsutton that in major cases dealing with unchecked exceptions is just a lot of boilerplate code unfortunately. 

But there were always two opposite camps of developers regarding checked exception and of course they are both right at some point
